### ISSUE TITLE
next/image not working when using in esm projects

### ISSUE THREAD

**Issue by yanush** [vercel/next.js#54777]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.17.1
      npm: 9.6.7
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.20-canary.12
      eslint-config-next: 13.4.19
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.2.2
    Next.js Config:
      output: N/A
```

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/yanush/nextjs_issue_20230830

### To Reproduce

- Create a new nextjs app:
`npx create-next-app@latest`

Using the following options: 
- TypeScript: No
- EsLint: Yes
- Tailwind CSS: No
- src/ directory: No
- App Router: Yes
- import Alias: No

- Edit package.json and add:
`"type": "module",`

Edit next.config.js file to match esm:
```
/** @type {import('next').NextConfig} */
const nextConfig = {}

export default nextConfig;
```

- try to run project using npm run dev

### Describe the Bug

I get these errors:
Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: object.

Check your code at page.js:19.
Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: object.

The errors referring to lines that contain _Image_ Tags

Notice:

- Using the legacy Image works fine (Changing page.js import to: `import Image from 'next/legacy/image'`)

- Changing the project to commonjs also seems to solve the issue

Both of these fixes are not ideal for me

### Expected Behavior

Should be working fine with the new image and don't throw errors

<sub>[NEXT-1774](https://linear.app/vercel/issue/NEXT-1774/nextimage-not-working-when-using-in-esm-projects)</sub>

**Comment by devjiwonchoi** [comment]
`type: module` is not implemented yet, so for now modify `next.config.js` as `next.config.mjs`.

**Comment by yanush** [comment]
It was working just fine in previous versions of next.js 13.
Was there any recent change that made it not to work again?

**Comment by devjiwonchoi** [comment]
@yanush Which specific version was working?

**Comment by yanush** [comment]
@devjiwonchoi version 13.3.4 with experimental appDir true was working fine

**Comment by yanush** [comment]
Any updates on this matter?
I hate having to use the legacy image just to keep using esm projects

**Comment by karlhorky** [comment]
Interesting, we have a working `"type": "module"` project here which also uses the `<Image />` component from `next/image`:

- https://github.com/upleveled/preflight-test-project-next-js-passing/blob/47c457e8cbeb12ea210f327733be4307822ac464/package.json#L5
- https://github.com/upleveled/preflight-test-project-next-js-passing/blob/47c457e8cbeb12ea210f327733be4307822ac464/next.config.js
- https://github.com/upleveled/preflight-test-project-next-js-passing/blob/47c457e8cbeb12ea210f327733be4307822ac464/app/page.tsx#L1

**Comment by karlhorky** [comment]
## Workaround

Change `page.js` to `page.tsx` - with TypeScript the `<Image />` component does not fail! 🤯

**Comment by karlhorky** [comment]
So I upgraded the original reproduction to `next@14.0.3-canary.5` (latest canary at time of writing) and can still confirm the bug:

1. `"type": "module"` in `package.json`
2. `app/page.js` (JavaScript, not TypeScript)
3. Usage of `<Image />` from `'next/image'`

Seems like a JS compiler bug?

Updated reproduction PR here:

https://github.com/yanush/nextjs_issue_20230830/pull/1

**Comment by karlhorky** [comment]
Ah interesting, since this `unstable_getImgProps` (introduced in #51205 by @styfle ) is also being returned, maybe this creates this problem with ESM:

```js
const Image = ImageModule.default;
// Image { unstable_getImgProps: [Getter], default: [Getter] }
```

**Comment by karlhorky** [comment]
Ah interesting, since this `unstable_getImgProps` (introduced in #51205 by @styfle ) is also being returned, maybe this creates this problem with ESM:

```js
console.log('Image', Image);
// Image { unstable_getImgProps: [Getter], default: [Getter] }
```

**Comment by yanush** [comment]
@karlhorky, Thanks for all your suggestions and workaround attempts!

**Comment by karlhorky** [comment]
Just thinking, one way to get compatibility here without a larger CJS -> ESM transition would be to \*also\* export a named export for these cases.

Then the following would "just work" in ESM with `"type": "module"` projects:

```js
import { Image } from 'next/image';
```

cc @feedthejim @shuding would the team be open for such a change? (backwards compatible)

cc @remcohaszing because we were talking about lacking ESM support in Next.js and [using named exports](https://twitter.com/remcohaszing/status/1718192390861648175)

**Comment by remcohaszing** [comment]
While I personally prefer named exports, I also believe there should be only one way to use an export. Adding a named export is a workaround. There is a deeper problem, this issue is just a symptom. Also it affects pretty much all Next.js exports, not just `next/image`.

A library (defined as something you import, e.g. `next`), should **not** rely on `esModuleInterop` if it’s compiled to CommonJS.

What really needs to happen, is that code like this:

```tsx
import React from 'react'

export function unstable_getImgProps() {
  // …
}

export interface ImageProps {
  // …
}

export default function Image(props: ImageProps): React.ReactElement {
  // …
}
```

needs to be transformed to `export =` syntax and a namespace.

```tsx
import * as React from 'react'

namespace Image {
  export function unstable_getImgProps() {
    // …
  }

  export interface ImageProps {
    // …
  }

  // This is for strict backwards compatibility. It’s not needed for typical Next.js users.
  export { Image as default }
}

function Image(props: ImageProps): React.ReactElement {
  // …
}

export = Image
```

Or go a step further and enable `verbatimModuleSyntax`:

```tsx
import React = require('react')

namespace Image {
  export function unstable_getImgProps() {
    // …
  }

  export interface ImageProps {
    // …
  }

  // This is for strict backwards compatibility. It’s not needed for typical Next.js users.
  export { Image as default }
}

function Image(props: ImageProps): React.ReactElement {
  // …
}

export = Image
```

---

P.s. If Vercel is invested in supporting ESM/TypeScript (which means fixing this), I am for hire to fix this on a freelance contract.

**Comment by karlhorky** [comment]
Would be amazing if full ESM support was fixed in Next.js 👍  Maybe Vercel would sponsor this (considerable) effort

Originally, `"type": "module"` support was added by @timneutkens in this PR:

- https://github.com/vercel/next.js/pull/33637

Basic support for ESM with TypeScript `'NodeNext'` or `'Node16'` module resolution added by @loettz here:

- https://github.com/vercel/next.js/pull/44177

But since then, there have been a number of issues, pull requests and discussions showing problems with ESM support (sometimes in combination with the TypeScript `'NodeNext'` or `'Node16'` module resolution), just a small sampling:

- https://github.com/vercel/next.js/issues/58509
- https://github.com/vercel/next.js/issues/34448
- https://github.com/vercel/next.js/issues/37419
- https://github.com/vercel/next.js/pull/34451
- https://github.com/vercel/next.js/pull/58380
- https://github.com/vercel/next.js/pull/50357
- https://github.com/vercel/next.js/issues/46678
- https://github.com/vercel/next.js/discussions/41189

cc @kachkaev @cseas because you have also been active in some issues and discussions

**Comment by karlhorky** [comment]
If this work to make the `next` package ESM-ready for most cases were to go ahead, I would suggest also including the following tools as part of the verification:

[Are The Types Wrong?](https://arethetypeswrong.github.io/?p=next%4014.0.3) by @andrewbranch

(results currently unavailable, seems like the tool cannot deal with `85784923 bytes gzipped` or maybe has a bug with unzipping)

![Screenshot 2023-11-19 at 12 12 32](https://github.com/vercel/next.js/assets/1935696/98770f3d-c471-4066-b3dd-8b2cdf02b37c)


[publint](https://publint.dev/next@14.0.3) by @bluwy

![Screenshot 2023-11-19 at 12 06 51](https://github.com/vercel/next.js/assets/1935696/96cdc030-89d2-4d4b-bfcc-a74c8e22a1c5)

**Comment by karlhorky** [comment]
Oh and it would be great if new end to end tests were added to verify continued ESM support as time goes on and new features get added.

Ideas for end to end tests:

1. A `"type": "module"` project mixing JavaScript and TypeScript with some commonly-used patterns such as importing `next/image`, `next/link`, including Tailwind CSS + PostCSS configuration
2. Usage of some static analysis tooling on published package artifacts such as the command line packages [`publint`](https://www.npmjs.com/package/publint) and [`@arethetypeswrong/cli`](https://github.com/arethetypeswrong/arethetypeswrong.github.io/tree/main/packages/cli)

**Comment by karlhorky** [comment]
@huozhi thanks for the fix in #59852

I can confirm that `next@14.0.5-canary.25` resolves the `Unsupported Server Component type` error, at least with `next/image` 🙌 :

---

Broken with `next@14.0.5-canary.24`

CodeSandbox demo: https://codesandbox.io/p/devbox/esm-js-import-bug-w-next-image-14-0-5-canary-24-773jvq?file=%2Fpnpm-lock.yaml%3A10%2C62

![Screenshot 2023-12-24 at 16 57 48](https://github.com/vercel/next.js/assets/1935696/ab07fda5-be21-442d-853f-3fc0c846c910)

---

Working with `next@14.0.5-canary.25`

CodeSandbox demo: https://codesandbox.io/p/devbox/esm-js-import-bug-w-next-image-14-0-5-canary-25-x58yz8?file=%2Fpnpm-lock.yaml%3A10%2C62

![Screenshot 2023-12-24 at 17 06 05](https://github.com/vercel/next.js/assets/1935696/cf79f14a-50e6-4551-9a0f-087c41f48c82)
